### PR TITLE
[core]refactor communication layer: PR1(Added Refactor Infra Only)

### PR DESCRIPTION
**Pull request by natureofnature**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

##  Purpose
Refactor the vLLM-Omni communication layer by moving all data-plane connector.put()/connector.get() calls from Orchestrator, OmniStage, and Scheduler into a unified OmniConnectorModelRunnerMixin at the Model Runner level, while keeping scheduling coordination logic (e.g., WAITING_FOR_CHUNK state) in the Scheduler.

-  Refer to RFC #1546 

This is the first PR for communication refactoring, which does not change existing workflow. The core files include Omni connector mixin and scheduler coordinator. 

## Architecture
architecture

## Status Summary
- [x] Main control-plane vs data-plane splitted.
- [x] `OmniConnectorModelRunnerMixin`  as the main communication hub for the refactored path.
- [x] Qwen2.5, Qwen 3 omni (aysnc & non async), Bagel works  
- [x] Batch and async chunk flows are mostly unified
- [x] chunk life cycle management in model runner to avoid chunk data transmission between model runner and scheduler
- [x] Legacy compatibility paths to be removed.
- [x] Fully absorb kv cache transfer into the mixin.
- [x] Heterogeneous rank support (kv cache transfer/full payload).
- [x] Heterogeneous rank support (async chunk).
- [x] Code Simplification and Reduction
- [x] Rebase with main

## Test Plan
```bash
pytest -s -v tests/e2e/ --ignore-glob='*expansion.py' -m "advanced_model" --run-level=advanced_model
```

## Test Result
Screenshot from 2026-03-29 10-13-23

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by lishunyang12 — COMMENTED** [review]
Left a few comments. This is a big refactor -- worth getting the threading right before it goes in.

File: vllm_omni/worker/omni_connector_model_runner_mixin.py [hunk 1]
```diff
@@ -0,0 +1,745 @@
...
+            with self._lock:
+                pending_ids = list(self._pending_load_reqs.keys())
+
+            if not pending_ids:
+                self._work_available.wait(timeout=0.01)
+                self._work_available.clear()
+                continue
+
+            for req_id in pending_ids:
+                if self._stop_event.is_set():
```
Comment (lishunyang12) [inline vllm_omni/worker/omni_connector_model_runner_mixin.py:1422]: `time.sleep(0.001)` after processing all pending IDs means this thread busy-spins at ~1kHz when there is work. The `_work_available.wait()` path only triggers when `pending_ids` is empty. Move the sleep inside the "no work" branch or replace it with an event wait after work is done too.

File: vllm_omni/worker/omni_connector_model_runner_mixin.py [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+    def _empty_output_with_connector_signals(self) -> Any:
+        """Return a minimal ModelRunnerOutput carrying pending connector signals.
+
+        Used by early-return paths (e.g. ``num_scheduled_tokens == 0``)
+        that still need to deliver ``omni_connector_output`` to the
+        Scheduler so that WAITING_FOR_INPUT / WAITING_FOR_CHUNK
+        transitions are not lost.
+        """
+        from copy import copy
+
```
Comment (lishunyang12) [inline vllm_omni/worker/omni_connector_model_runner_mixin.py:1269]: `copy()` is a shallow copy -- if `EMPTY_MODEL_RUNNER_OUTPUT` has any mutable default fields (lists, dicts), they will be shared between the copy and the singleton. This is fine today since you only set `omni_connector_output`, but it is fragile. Consider `copy.deepcopy()` or constructing a fresh `OmniModelRunnerOutput` directly.

File: vllm_omni/worker/omni_connector_model_runner_mixin.py [hunk 3]
```diff
@@ -0,0 +1,745 @@
...
+        self._put_req_chunk: dict[str, int] = defaultdict(int)
+        self._get_req_chunk: dict[str, int] = defaultdict(int)
+        self._request_payload: dict[str, dict[str, Any]] = {}
+        self._code_prompt_token_ids: dict[str, list[list[int]]] = defaultdict(list)
+        self._request_ids_mapping: dict[str, str] = {}
+
+        # -- async I/O state (shared by chunk + batch) --
+        self._pending_load_reqs: dict[str, Any] = {}
+        self._finished_load_reqs: set[str] = set()
+        self._pending_save_reqs: dict[str, deque] = {}
```
Comment (lishunyang12) [inline vllm_omni/worker/omni_connector_model_runner_mixin.py:128]: `_finished_save_reqs` is initialized here but never read anywhere in the class. Dead state.

File: tests/e2e/online_serving/test_refactored_communication.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+    def test_full_cycle(self):
+        from vllm.v1.request import RequestStatus
+
+        coord = ChunkSchedulingCoordinator(scheduler_max_num_seqs=10, stage_id=1)
+
+        req = _make_request("r1", status=RequestStatus.WAITING)
+
+        class MockQueue:
+            def __init__(self, items=None):
```
Comment (lishunyang12) [inline tests/e2e/online_serving/test_refactored_communication.py]: This `MockQueue` is a copy of the module-level one already defined at line 56. Use the existing one.

File: tests/core/sched/test_chunk_scheduling_coordinator.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+# ------------------------------------------------------------------ #
+#  Mock helpers
+# ------------------------------------------------------------------ #
+
+
+class _RequestStatus:
+    WAITING = "waiting"
+    RUNNING = "running"
+    WAITING_FOR_CHUNK = "waiting_for_chunk"
+    WAITING_FOR_INPUT = "waiting_for_input"
```
Comment (R2-Y) [inline tests/core/sched/test_chunk_scheduling_coordinator.py:29]: why we need two request status WAITING_FOR_CHUNK &  WAITING_FOR_INPUT? CHUNK is one of the input?

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 6]
```diff
@@ -53,8 +67,34 @@ def schedule(self) -> SchedulerOutput:
         # Temporary queue: preserve waiting order, do not disturb non-diffusion requests
         skipped_waiting_requests = create_request_queue(self.policy)
         req_index = 0
-        if self.chunk_transfer_adapter:
-            self.chunk_transfer_adapter.process_pending_chunks(self.waiting, self.running)
+        if self.chunk_coordinator:
+            oco = self._latest_omni_connector_output
```
Comment (R2-Y) [inline vllm_omni/core/sched/omni_generation_scheduler.py]: the variable names are too vague

File: vllm_omni/core/sched/omni_scheduling_coordinator.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+        # Requests waiting for batch stage input (WAITING_FOR_INPUT).
+        self._waiting_for_input: deque[Any] = deque()
+        self.pending_input_registrations: list[Any] = []
+
+    # ------------------------------------------------------------------ #
+    #  Core scheduling methods
+    # ------------------------------------------------------------------ #
+
+    def process_pending_chunks(
```
Comment (R2-Y) [inline vllm_omni/core/sched/omni_scheduling_coordinator.py:64]: Is it possible to put the processing of Async Chunk and non-async chunk code blocks in the same function? Currently, some processing seems replicated in process_pending_chunks & process_pending_batch_inputs. Could we use a single queue, `waiting_for_input_waiting`, to handle both `waiting_for_input` and `waiting_for_chunk_waiting` simultaneously, and only hold one request status (e.g., `WAITING_FOR_INPUT`) for both asynchronous and non-asynchronous chunk modes?

File: vllm_omni/core/sched/output.py [hunk 8]
```diff
@@ -78,3 +78,9 @@ class OmniSchedulerOutput(SchedulerOutput):
     """Scheduler output with omni-specific transfer metadata."""
 
     finished_requests_needing_kv_transfer: dict[str, dict] = field(default_factory=dict)
+    # Requests that need to be registered for chunk recv by the Model Runner's
+    # background thread. Populated by ChunkSchedulingCoordinator.
+    pending_chunk_registrations: list = field(default_factory=list)
```
Comment (R2-Y) [inline vllm_omni/core/sched/output.py]: same as above. can we keep only one for registration?

File: vllm_omni/entrypoints/async_omni.py [hunk 9]
```diff
@@ -456,46 +446,34 @@ async def _process_sequential_results(
...
-                logger.debug(f"[{self._name}] Forwarded request {request_id} to stage-{next_stage_id}")
-            else:
+                except Exception:
+                    logger.warning("compute_talker_prompt_ids_length failed, falling back to len()", exc_info=True)
+                    next_prompt_len = max(1, len(prompt_token_ids)) if prompt_token_ids else 1
+                seed_input = copy.deepcopy(prompt) if isinstance(prompt, dict) else {"prompt_token_ids": []}
+                seed_input["prompt_token_ids"] = [0] * next_prompt_len
+                seed_input["multi_modal_data"] = seed_input["mm_processor_kwargs"] = None
+                for ds_stage_id in range(1, len(self.stage_list)):
+                    sp_ds = sampling_params_list[ds_stage_id]
```
Comment (R2-Y) [inline vllm_omni/entrypoints/async_omni.py]: the variable name is too vague

**Comment by lishunyang12** [comment]
Hey @natureofnature — I see two new commits since my review, but my 4 inline comments from 2/28 still seem open (busy-spin in recv_loop, shallow copy, dead _finished_save_reqs, duplicate MockQueue in e2e test). Could you take a look when you get a chance?

**Comment by natureofnature** [comment]
> Hey @natureofnature — I see two new commits since my review, but my 4 inline comments from 2/28 still seem open (busy-spin in recv_loop, shallow copy, dead _finished_save_reqs, duplicate MockQueue in e2e test). Could you take a look when you get a chance?

I'm still working on it, and will fix them once the basic workflow functions well. @lishunyang12

**Comment by natureofnature** [comment]
@codex review

**Comment by natureofnature** [comment]
@codex review

**Comment by natureofnature** [comment]
@tzhouam  @divyanshsinghvi PTAL

**Review by hsliuustc0106 — DISMISSED** [review]
## Gate Status: ❌ BLOCKED

| Check | Status |
|-------|--------|
| DCO | ✅ |
| Docs | ✅ |
| **Mergeable** | ❌ CONFLICTING |

---

### Prior Feedback Unaddressed (2/28)

1. **Dead state**: `_finished_save_reqs` initialized but never read
2. **Silent connector failure**: Returns `None` instead of failing fast
3. **Busy-spin**: `time.sleep(0.001)` placement still suboptimal
4. **Duplicate MockQueue**: Same class in two test files

---

### Test Evidence Missing

Claims "Qwen2.5, Qwen 3 omni, Bagel works" but provides no test commands or output. Please add reproduction steps.

---

### Minor: MRO Inconsistency

```python
# gpu_ar_model_runner.py:154
class GPUARModelRunner(OmniGPUModelRunner, OmniConnectorModelRunnerMixin):  # ✓

# gpu_generation_model_runner.py:45
class GPUGenerationModelRunner(OmniConnectorModelRunnerMixin, OmniGPUModelRunner):  # ✗
```

---

Please resolve conflicts and address the 4 prior items.

File: vllm_omni/entrypoints/omni.py [hunk 10]
```diff
@@ -1063,47 +1066,42 @@ def _run_generation(
...
+                                exc_info=True,
+                            )
+                            next_prompt_len = max(1, len(prompt_token_ids)) if prompt_token_ids else 1
+                        seed_input = {
+                            "prompt_token_ids": [0] * next_prompt_len,
+                            "multi_modal_data": None,
+                            "mm_processor_kwargs": None,
+                        }
+                        for ds_stage_id in range(1, len(self.stage_list)):
+                            sp_ds = sampling_params_list[ds_stage_id]
```
[the remark this answers is not in the document]
Comment (natureofnature) [reply vllm_omni/entrypoints/omni.py]: resolved

**Comment by natureofnature** [comment]
@codex review

**Comment by natureofnature** [comment]
> ## Gate Status: ❌ BLOCKED
> Check	Status
> DCO	✅
> Docs	✅
> **Mergeable**	❌ CONFLICTING
> ### Prior Feedback Unaddressed (2/28)
> 1. **Dead state**: `_finished_save_reqs` initialized but never read
> 2. **Silent connector failure**: Returns `None` instead of failing fast
> 3. **Busy-spin**: `time.sleep(0.001)` placement still suboptimal
> 4. **Duplicate MockQueue**: Same class in two test files
> 
> ### Test Evidence Missing
> Claims "Qwen2.5, Qwen 3 omni, Bagel works" but provides no test commands or output. Please add reproduction steps.
> 
> ### Minor: MRO Inconsistency
> ```python
> # gpu_ar_model_runner.py:154
> class GPUARModelRunner(OmniGPUModelRunner, OmniConnectorModelRunnerMixin):  # ✓
> 
> # gpu_generation_model_runner.py:45
> class GPUGenerationModelRunner(OmniConnectorModelRunnerMixin, OmniGPUModelRunner):  # ✗
> ```
> 
> Please resolve conflicts and address the 4 prior items.

resolved

File: vllm_omni/entrypoints/talker_prompt_utils.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/talker_prompt_utils.py]: talker is too model specific, which should not be listed here

**Review by hsliuustc0106 — COMMENTED** [review]
this PR to too large, can we test this in one model intensively before moving to all models?

**Comment by natureofnature** [comment]
> this PR to too large, can we test this in one model intensively before moving to all models?
I split the pr to two parts, this is the first part, which does not change existing workflow. @hsliuustc0106

File: vllm_omni/worker/payload_span.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hsliuustc0106) [inline vllm_omni/worker/payload_span.py:1]: why we put it here under vllm_omni/?

File: vllm_omni/worker/omni_connector_model_runner_mixin.py [hunk 13]
```diff
@@ -0,0 +1,2094 @@
...
+from typing import TYPE_CHECKING, Any
+
+import torch
+from vllm.distributed.parallel_state import get_tp_group
+from vllm.logger import init_logger
+
+from vllm_omni.distributed.omni_connectors.factory import OmniConnectorFactory
+from vllm_omni.distributed.omni_connectors.utils.config import ConnectorSpec
+from vllm_omni.outputs import OmniConnectorOutput
+from vllm_omni.payload_span import (
```
Comment (hsliuustc0106) [inline vllm_omni/worker/omni_connector_model_runner_mixin.py]: should we move this to vllm_omni/worker?

File: vllm_omni/core/sched/omni_scheduling_coordinator.py [hunk 14]
```diff
@@ -0,0 +1,380 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (hsliuustc0106) [inline vllm_omni/core/sched/omni_scheduling_coordinator.py:1]: @amy-why-3459 PTAL

File: vllm_omni/worker/payload_span.py (hunk 12 above)
Comment (natureofnature) [reply vllm_omni/worker/payload_span.py:1]: I moved it into vllm_omni/worker. It's indeed should not be there.

File: vllm_omni/worker/omni_connector_model_runner_mixin.py (hunk 13 above)
Comment (natureofnature) [reply vllm_omni/worker/omni_connector_model_runner_mixin.py]: moved

**Comment by hsliuustc0106** [comment]
merge ci passed but nightly ci failed

**Comment by natureofnature** [comment]
> merge ci passed but nightly ci failed

I fixed one issue that caused by this PR. And it seems current nightly CI problems (https://buildkite.com/vllm/vllm-omni/builds/6413/steps/canvas) are not related to this PR. @hsliuustc0106
